pkg/inst/worker/daemon.R

```diff
@@ -0,0 +1,57 @@
+# Worker daemon
+
+rLibDir <- Sys.getenv("SPARKR_RLIBDIR")
+script <- paste(rLibDir, "SparkR/worker/worker.R", sep = "/")
+
+# preload SparkR package, speedup worker
+.libPaths(c(rLibDir, .libPaths()))
+suppressPackageStartupMessages(library(SparkR))
+
+port <- as.integer(Sys.getenv("SPARKR_WORKER_PORT"))
+inputCon <- socketConnection(port = port, open = "rb", blocking = TRUE, timeout = 3600)
+
+# Read from connection, retrying to read exactly 'size' bytes
+readBinFull <- function(con, size) {
+  # readBin() could be interruptted by signal, we have no way to tell
+  # if it's interruptted or the socket is closed, so we retry at most
+  # 20 times, to avoid the deadloop if the socket is closed
+  c <- 1
```

Comment:
Yeah I don't know why exactly readBin behaves the way it does -- I think what happens is that calling `read` on a socket is best effort and returns how many ever bytes are available[2]. In fact I can't find any code in R [1] which ensures that the read is retried until all the required bytes are read. I think we don't run into this for larger message sizes, but in unit tests with only 4 bytes being sent very frequently we tend to run into this. 

I'm actually thinking this patterns needs to become more widespread (i.e. for the SparkR JVM backend as well ?)

[1] https://github.com/wch/r-source/blob/trunk/src/main/connections.c#L3797
[2] http://beej.us/guide/bgnet/output/html/multipage/recvman.html